File: examples/java/ExampleReduceBenchmark.java
import java.io.IOException;

import yacx.Executor;
import yacx.Executor.KernelArgCreator;
import yacx.LongArg;
import yacx.IntArg;
import yacx.KernelArg;
import yacx.Options;

public class ExampleReduceBenchmark {

	public static void main(String[] args) throws IOException {
		// Load library
		Executor.loadLibrary();

		KernelArgCreator<Integer> creator1 = new Executor.KernelArgCreator<Integer>() {

			@Override
			public int getGrid0(Integer dataLength) {
				int numThreads = getBlock0(0);
				return Math.min((dataLength + numThreads - 1) / numThreads, 1024);
			}

			@Override
			public int getBlock0(Integer dataLength) {
				return 512;
			}

			@Override
			public KernelArg[] createArgs(Integer dataLength) {
				long[] in = new long[dataLength];
				for (int i = 0; i < in.length; i++) {
					in[i] = i;
				}

				LongArg inArg = LongArg.create(in);
				LongArg outArg = LongArg.createOutput(dataLength);
				KernelArg nArg = IntArg.createValue(dataLength);

				return new KernelArg[] { inArg, outArg, nArg };
			}
		};

		KernelArgCreator<Integer> creator2 = new Executor.KernelArgCreator<Integer>() {

			@Override
			public int getGrid0(Integer dataLength) {
				return 1;
			}

			@Override
			public int getBlock0(Integer dataLength) {
				return 1024;
			}

			@Override
			public KernelArg[] createArgs(Integer dataLength) {
				int blocks = Math.min((dataLength + 512 - 1) / 512, 1024);
				long[] in = new long[blocks];
				for (int i = 0; i < in.length; i++) {
					in[i] = i;
				}

				LongArg inArg = LongArg.create(in);
				LongArg outArg = LongArg.createOutput(blocks);
				KernelArg nArg = IntArg.createValue(blocks);

				return new KernelArg[] { inArg, outArg, nArg };
			}
		};

		Integer[] dataSizes = new Integer[] {1024, 2048, 4096, 8192, 256 * 1024*1024};

		// Options
		Options options = Options.createOptions();

		// Warm up
		Executor.benchmark("device_reduce", options, 30, creator1, 1024*1024);

		// Benchmark Reduce-Kernel
		Executor.BenchmarkResult<Integer> result = Executor.benchmark("device_reduce", options, 50, creator1, dataSizes);

		// Simulate second kernel call
		Executor.BenchmarkResult<Integer> result2 = Executor.benchmark("device_reduce", options, 50, creator2, dataSizes);

		// Add the average times of the second benchmark to the first
		Executor.BenchmarkResult<Integer> sum = result.addBenchmarkResult(result2);

		// Print out the final benchmark result
		System.out.println(sum);
	}
}

File: examples/scala/ExampleReduceBenchmark.scala
import java.io.IOException
import yacx.Executor
import yacx.Executor.KernelArgCreator
import yacx.LongArg
import yacx.IntArg
import yacx.KernelArg
import yacx.Options

object ExampleReduceBenchmark {
  private val KB = 1024
  private val MB = 1024 * 1024

  def main(args: Array[Nothing]): Unit = {
    // Load library
    Executor.loadLibrary()

    val creator1 = new Executor.KernelArgCreator() {
      override def getDataLength(dataSizeBytes: Long): Int = (dataSizeBytes / LongArg.SIZE_BYTES).toInt

      override def getGrid0(dataLength: Int): Int = {
        val numThreads = getBlock0(0)
        Math.min((dataLength + numThreads - 1) / numThreads, 1024)
      }

      override def getBlock0(dataLength: Int) = 512

      override def createArgs(dataLength: Int): Array[KernelArg] = {
        val in = new Array[Long](dataLength)
        for (i <- 0 until in.length) {
          in(i) = i
        }
        val inArg = LongArg.create(in)
        val outArg = LongArg.createOutput(dataLength)
        val nArg = IntArg.createValue(dataLength)
        Array[KernelArg](inArg, outArg, nArg)
      }
    }
    val creator2 = new Executor.KernelArgCreator() {
      override def getDataLength(dataSizeBytes: Long): Int = (dataSizeBytes / LongArg.SIZE_BYTES).toInt

      override def getGrid0(dataLength: Int) = 1

      override def getBlock0(dataLength: Int) = 1024

      override def createArgs(dataLength: Int): Array[KernelArg] = {
        val blocks = Math.min((dataLength + 512 - 1) / 512, 1024)
        val in = new Array[Long](blocks)
        for (i <- 0 until in.length) {
          in(i) = i
        }
        val inArg = LongArg.create(in)
        val outArg = LongArg.createOutput(blocks)
        val nArg = IntArg.createValue(blocks)
        Array[KernelArg](inArg, outArg, nArg)
      }
    }
    val dataSizes = Array[Long](1 * KB, 4 * KB, 16 * KB, 64 * KB, 256 * KB, 1 * MB, 4 * MB, 16 * MB, 64 * MB, 256 * MB, 1024 * MB)

    // Options
    val options = Options.createOptions

    // Warm up
    Executor.benchmark("device_reduce", options, 30, creator1, 256 * MB)

    // Benchmark Reduce-Kernel
    val result = Executor.benchmark("device_reduce", options, 50, creator1, dataSizes)

    // Simulate second kernel call
    val result2 = Executor.benchmark("device_reduce", options, 50, creator2, dataSizes)

    // Add the average times of the second benchmark to the first
    val sum = result.addBenchmarkResult(result2)

    // Print out the final benchmark result
    println(sum)
  }
}